File: backend-nodejs/src/otel-config.ts
/**
 * OpenTelemetry Configuration - Enterprise observability for Node.js
 *
 * Provides:
 * - Automatic Express instrumentation
 * - OTLP export to Jaeger/Honeycomb/Datadog
 * - Context propagation
 * - Custom span creation
 *
 * Usage:
 *   // MUST be imported first, before other modules
 *   import { setupOtel, getTracer } from './otel-config';
 *   setupOtel();
 *
 *   // Create custom spans
 *   const tracer = getTracer('my-module');
 *   const span = tracer.startSpan('my_operation');
 *   span.setAttribute('user_id', userId);
 *   span.end();
 *
 * Configuration:
 *   OTEL_ENABLED=true
 *   OTEL_SERVICE_NAME=beyondcloud-api
 *   OTEL_EXPORTER=otlp|console
 *   OTEL_ENDPOINT=http://localhost:4317
 *   OTEL_HEADERS=x-honeycomb-team=your-key
 */

// Types for when OTel is not installed
interface Span {
    setAttribute(key: string, value: string | number | boolean): void;
    setStatus(status: { code: number; message?: string }): void;
    addEvent(name: string, attributes?: Record<string, any>): void;
    end(): void;
}

interface Tracer {
    startSpan(name: string, options?: any): Span;
}

let _otelAvailable = false;
let _api: any = null;
let _sdk: any = null;

// Try to import OpenTelemetry packages
async function loadOtel() {
    try {
        _api = await import('@opentelemetry/api');
        const { NodeTracerProvider } = await import('@opentelemetry/sdk-trace-node');
        const { BatchSpanProcessor, ConsoleSpanExporter } = await import('@opentelemetry/sdk-trace-base');
        const { Resource } = await import('@opentelemetry/resources');
        const { SEMRESATTRS_SERVICE_NAME, SEMRESATTRS_SERVICE_VERSION } = await import('@opentelemetry/semantic-conventions');
        const { OTLPTraceExporter } = await import('@opentelemetry/exporter-trace-otlp-grpc');
        const { ExpressInstrumentation } = await import('@opentelemetry/instrumentation-express');
        const { HttpInstrumentation } = await import('@opentelemetry/instrumentation-http');
        const { registerInstrumentations } = await import('@opentelemetry/instrumentation');

        _sdk = {
            NodeTracerProvider,
            BatchSpanProcessor,
            ConsoleSpanExporter,
            Resource,
            SEMRESATTRS_SERVICE_NAME,
            SEMRESATTRS_SERVICE_VERSION,
            OTLPTraceExporter,
            ExpressInstrumentation,
            HttpInstrumentation,
            registerInstrumentations,
        };
        _otelAvailable = true;
        return true;
    } catch (error) {
        console.warn('OpenTelemetry packages not installed. Install with:');
        console.warn('npm install @opentelemetry/api @opentelemetry/sdk-trace-node @opentelemetry/exporter-trace-otlp-grpc @opentelemetry/instrumentation-express @opentelemetry/instrumentation-http');
        return false;
    }
}

/**
 * Setup OpenTelemetry tracing
 * MUST be called before importing Express or other instrumented modules
 */
export async function setupOtel(serviceName?: string, serviceVersion: string = '1.0.0'): Promise<boolean> {
    const enabled = process.env.OTEL_ENABLED?.toLowerCase() === 'true';
    if (!enabled) {
        console.log('OpenTelemetry disabled (set OTEL_ENABLED=true to enable)');
        return false;
    }

    const loaded = await loadOtel();
    if (!loaded) {
        return false;
    }

    const name = serviceName || process.env.OTEL_SERVICE_NAME || 'beyondcloud-api';
    const environment = process.env.NODE_ENV || 'development';

    // Create resource
    const resource = new _sdk.Resource({
        [_sdk.SEMRESATTRS_SERVICE_NAME]: name,
        [_sdk.SEMRESATTRS_SERVICE_VERSION]: serviceVersion,
        'deployment.environment': environment,
    });

    // Create provider
    const provider = new _sdk.NodeTracerProvider({ resource });

    // Configure exporter
    const exporterType = process.env.OTEL_EXPORTER?.toLowerCase() || 'otlp';

    if (exporterType === 'console') {
        provider.addSpanProcessor(new _sdk.BatchSpanProcessor(new _sdk.ConsoleSpanExporter()));
        console.log('Using Console span exporter');
    } else {
        const endpoint = process.env.OTEL_ENDPOINT || 'http://localhost:4317';
        const headers = parseHeaders(process.env.OTEL_HEADERS || '');

        const exporter = new _sdk.OTLPTraceExporter({
            url: endpoint,
            headers,
        });
        provider.addSpanProcessor(new _sdk.BatchSpanProcessor(exporter));
        console.log(`Using OTLP exporter: ${endpoint}`);
    }

    // Register provider
    provider.register();

    // Auto-instrument Express and HTTP
    _sdk.registerInstrumentations({
        instrumentations: [
            new _sdk.HttpInstrumentation(),
            new _sdk.ExpressInstrumentation(),
        ],
    });

    console.log(`OpenTelemetry initialized for ${name}`);
    return true;
}

function parseHeaders(headersStr: string): Record<string, string> {
    if (!headersStr) return {};

    const headers: Record<string, string> = {};
    for (const item of headersStr.split(',')) {
        const [key, value] = item.split('=');
        if (key && value) {
            headers[key.trim()] = value.trim();
        }
    }
    return headers;
}

/**
 * Get a tracer instance for creating custom spans
 */
export function getTracer(name: string = 'beyondcloud'): Tracer {
    if (!_otelAvailable || !_api) {
        return new NoOpTracer();
    }
    return _api.trace.getTracer(name);
}

/**
 * Get the current active span
 */
export function getCurrentSpan(): Span | undefined {
    if (!_otelAvailable || !_api) {
        return undefined;
    }
    return _api.trace.getActiveSpan();
}

/**
 * Get trace context for correlation
 */
export function getTraceContext(): { trace_id?: string; span_id?: string } {
    if (!_otelAvailable || !_api) {
        return {};
    }

    const span = _api.trace.getActiveSpan();
    if (!span) return {};

    const ctx = span.spanContext();
    return {
        trace_id: ctx.traceId,
        span_id: ctx.spanId,
    };
}

// No-op implementations for when OTel is not available
class NoOpSpan implements Span {
    setAttribute(_key: string, _value: string | number | boolean): void { }
    setStatus(_status: { code: number; message?: string }): void { }
    addEvent(_name: string, _attributes?: Record<string, any>): void { }
    end(): void { }
}

class NoOpTracer implements Tracer {
    startSpan(_name: string, _options?: any): Span {
        return new NoOpSpan();
    }
}

/**
 * Decorator-style function for tracing async operations
 */
export function traced<T>(
    name: string,
    fn: () => Promise<T>,
    attributes?: Record<string, string | number | boolean>
): Promise<T> {
    const tracer = getTracer();
    const span = tracer.startSpan(name);

    if (attributes) {
        for (const [key, value] of Object.entries(attributes)) {
            span.setAttribute(key, value);
        }
    }

    return fn()
        .then((result) => {
            span.end();
            return result;
        })
        .catch((error) => {
            span.setAttribute('error', true);
            span.setAttribute('error.message', error.message);
            span.end();
            throw error;
        });
}

File: backend-python/app/otel_config.py
"""
OpenTelemetry Configuration - Enterprise observability

Provides:
- Automatic FastAPI instrumentation
- OTLP export to Jaeger/Honeycomb/Datadog
- Context propagation
- Custom span creation

Usage:
    from app.otel_config import setup_otel, get_tracer
    
    # In main.py startup
    setup_otel(app)
    
    # Create custom spans
    tracer = get_tracer(__name__)
    with tracer.start_as_current_span("my_operation") as span:
        span.set_attribute("user_id", user_id)
        # ... do work

Configuration (environment variables):
    OTEL_ENABLED=true
    OTEL_SERVICE_NAME=beyondcloud-api
    OTEL_EXPORTER=otlp|jaeger|console
    OTEL_ENDPOINT=http://localhost:4317 (for OTLP)
    OTEL_HEADERS=x-honeycomb-team=your-key (for Honeycomb)
"""
import os
import logging
from typing import Optional
from contextlib import contextmanager

logger = logging.getLogger(__name__)

# Check if OpenTelemetry is available
_otel_available = False
try:
    from opentelemetry import trace
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
    from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION
    from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
    from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
    _otel_available = True
except ImportError:
    logger.warning("OpenTelemetry packages not installed. Run: pip install opentelemetry-api opentelemetry-sdk opentelemetry-instrumentation-fastapi opentelemetry-exporter-otlp")


def setup_otel(app=None, service_name: str = None, service_version: str = "1.0.0"):
    """
    Setup OpenTelemetry tracing.
    
    Args:
        app: FastAPI application instance (for auto-instrumentation)
        service_name: Service name for traces (default: OTEL_SERVICE_NAME or "beyondcloud-api")
        service_version: Service version for traces
    """
    if not _otel_available:
        logger.warning("OpenTelemetry not available, skipping setup")
        return
    
    enabled = os.getenv("OTEL_ENABLED", "false").lower() == "true"
    if not enabled:
        logger.info("OpenTelemetry disabled (set OTEL_ENABLED=true to enable)")
        return
    
    # Service info
    service_name = service_name or os.getenv("OTEL_SERVICE_NAME", "beyondcloud-api")
    
    # Create resource with service info
    resource = Resource.create({
        SERVICE_NAME: service_name,
        SERVICE_VERSION: service_version,
        "deployment.environment": os.getenv("ENVIRONMENT", "development"),
    })
    
    # Create tracer provider
    provider = TracerProvider(resource=resource)
    
    # Configure exporter based on environment
    exporter_type = os.getenv("OTEL_EXPORTER", "otlp").lower()
    
    if exporter_type == "console":
        # Console exporter for development/debugging
        exporter = ConsoleSpanExporter()
        logger.info("Using Console span exporter")
    else:
        # OTLP exporter (works with Jaeger, Honeycomb, Datadog, etc.)
        endpoint = os.getenv("OTEL_ENDPOINT", "http://localhost:4317")
        headers = _parse_headers(os.getenv("OTEL_HEADERS", ""))
        
        exporter = OTLPSpanExporter(
            endpoint=endpoint,
            headers=headers,
            insecure=endpoint.startswith("http://"),
        )
        logger.info(f"Using OTLP exporter: {endpoint}")
    
    # Add batch processor for efficiency
    provider.add_span_processor(BatchSpanProcessor(exporter))
    
    # Set as global tracer provider
    trace.set_tracer_provider(provider)
    
    # Auto-instrument FastAPI if app provided
    if app is not None:
        FastAPIInstrumentor.instrument_app(app)
        logger.info("FastAPI auto-instrumented with OpenTelemetry")
    
    logger.info(f"OpenTelemetry initialized for {service_name}")


def _parse_headers(headers_str: str) -> dict:
    """Parse headers from comma-separated key=value string"""
    if not headers_str:
        return {}
    
    headers = {}
    for item in headers_str.split(","):
        if "=" in item:
            key, value = item.split("=", 1)
            headers[key.strip()] = value.strip()
    return headers


def get_tracer(name: str = __name__):
    """
    Get a tracer instance for creating custom spans.
    
    Usage:
        tracer = get_tracer(__name__)
        with tracer.start_as_current_span("operation") as span:
            span.set_attribute("key", "value")
    """
    if not _otel_available:
        return _NoOpTracer()
    
    return trace.get_tracer(name)


def get_current_span():
    """Get the current active span"""
    if not _otel_available:
        return _NoOpSpan()
    return trace.get_current_span()


def get_trace_context() -> dict:
    """
    Get current trace context for correlation.
    Returns dict with trace_id and span_id if available.
    """
    if not _otel_available:
        return {}
    
    span = trace.get_current_span()
    ctx = span.get_span_context()
    
    if ctx.is_valid:
        return {
            "trace_id": format(ctx.trace_id, "032x"),
            "span_id": format(ctx.span_id, "016x"),
        }
    return {}


class _NoOpSpan:
    """No-op span for when OTel is not available"""
    def set_attribute(self, key, value): pass
    def set_status(self, status): pass
    def add_event(self, name, attributes=None): pass
    def end(self): pass
    def __enter__(self): return self
    def __exit__(self, *args): pass


class _NoOpTracer:
    """No-op tracer for when OTel is not available"""
    @contextmanager
    def start_as_current_span(self, name, **kwargs):
        yield _NoOpSpan()
    
    def start_span(self, name, **kwargs):
        return _NoOpSpan()


# =============================================================================
# Convenience decorators
# =============================================================================

def traced(name: str = None, attributes: dict = None):
    """
    Decorator to trace a function.
    
    Usage:
        @traced("my_function")
        async def my_function():
            ...
    """
    def decorator(func):
        import functools
        import asyncio
        
        span_name = name or func.__name__
        
        @functools.wraps(func)
        async def async_wrapper(*args, **kwargs):
            tracer = get_tracer(func.__module__)
            with tracer.start_as_current_span(span_name) as span:
                if attributes:
                    for k, v in attributes.items():
                        span.set_attribute(k, v)
                try:
                    return await func(*args, **kwargs)
                except Exception as e:
                    span.set_attribute("error", True)
                    span.set_attribute("error.message", str(e))
                    raise
        
        @functools.wraps(func)
        def sync_wrapper(*args, **kwargs):
            tracer = get_tracer(func.__module__)
            with tracer.start_as_current_span(span_name) as span:
                if attributes:
                    for k, v in attributes.items():
                        span.set_attribute(k, v)
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    span.set_attribute("error", True)
                    span.set_attribute("error.message", str(e))
                    raise
        
        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        return sync_wrapper
    
    return decorator
